### ISSUE TITLE
Remove external setup method for non-Custom build type

### ISSUE THREAD

**Issue by dcoutts** [haskell/cabal#2889]
Currently the SetupWrapper module has quite a lot of complexity due to supporting the 'externalSetupMethod' for the case of BuildType Simple/Configure/Make. In particular it has a setup exe cache that is only used in this case.

We can simplify this. Here's the current logic:

    determineSetupMethod options buildType'
      | buildType' == Custom             = externalSetupMethod
      | not (cabalVersion `withinRange`
             useCabalVersion options)    = externalSetupMethod
      | isJust (useLoggingHandle options)
        -- Forcing is done to use an external process e.g. due to parallel
        -- build concerns.
     || forceExternalSetupMethod options = selfExecSetupMethod
      | otherwise                        = internalSetupMethod

If we call these cases 1-4, my claim is that case 2 is just not necessary. In practice there is no sensible reason for it to happen and so no need to support it. The only case in which we need the externalSetupMethod is for custom Setup.hs, and it doesn't need any setup exe cache.

Why do we have case 2? There's two reasons, one historical and one unnecessary. The historical reason is that there used not to be a case 3, and so case 2 was needed for that. That is, we didn't used to have the 'selfExecSetupMethod' where cabal invokes itself to act as the 'Setup.hs' CLI, so we often had to compile a perfectly ordinary 'Setup.hs' scripts with various versions of the Cabal lib that we could find lying around. So that's the historical reason.

The other reason is that the 'useCabalVersion options' thing. But where does this come from and what values can it take? Internally within the Cabal code we only ever pass in minimum bounds for 'useCabalVersion'. So that never excludes the "current" version of Cabal that cabal-install was built with. The only external source for this is a command line flag `--cabal-lib-version` which is more or less just a debugging or workaround flag. We don't actually need it and it should not hold us back from making things simpler and more robust. For example, we could have it only apply to 'Custom' build type.

So, the task here is to abolish case 2 above. Get rid of 'useCabalVersion' or at least restrict its used to 'build-type: Custom'. And then get rid of the setup exe cache, since it's no longer needed.

**Comment by 23Skidoo** [comment]
I think that this logic can fail in the case of `build-type: Simple` and `cabal-version: > self-exec one`. In theory this can happen when you're using an old `cabal-install` exe with a new GHC and try to install a package that specifies a more recent `cabal-version`.

We can of course just print an error and refuse to proceed when the installed Cabal is more recent than the one we were compiled with, since old versions of `cabal-install` often don't work with new versions of GHC anyway.

**Comment by dcoutts** [comment]
I thought that cabal-install wouldn't even try to install a package where the .cabal file specifies `cabal-version` greater than what it iself understands (ie the Cabal lib it was built with). But apparently I'm wrong about that, it will. It's a bit of a gamble, we don't know if cabal-install can understand the any possibly new features, though I guess typically it'll be ok.

Ok, so the only case then is when it's a newer version than the self-exec one. Does this relatively rare case still justify the setup exe cache?

**Comment by 23Skidoo** [comment]
> Does this relatively rare case still justify the setup exe cache?

I'd be happy to get rid of it -- as I said, we can change `cabal-install` to exit with an error in this case (or at least print a warning).

**Comment by dcoutts** [comment]
Oh we could still keep the feature to allow an older cabal-install to call a newer Cabal lib by compiling a Setup.hs. But we could do that in the simpler old-fashioned way of just building the Setup in the dist tree like we do for custom Setup.hs. That way we can get rid of the cached setup code and complexity.

**Comment by 23Skidoo** [comment]
>  we could do that in the simpler old-fashioned way of just building the Setup in the dist tree like we do for custom Setup.hs.

So... this means that we're not removing the case 2 above, just the setup exe cache logic?

**Comment by dcoutts** [comment]
Yeah, I guess so. It's less of a reduction in complexity than I was hoping for, but I'd not considered that old cabal-install can talk to new Cabal lib for build-type Simple.

**Comment by 23Skidoo** [comment]
> old cabal-install can talk to new Cabal lib for build-type Simple

This can happen for all build types, I think.
